### PR TITLE
Rays sync ch

### PR DESCRIPTION
**Pull request by mingwandroid**
This replaces the previous PR and incorporates the work done by @ChristopherHogan in https://github.com/conda/conda-build/pull/337

### CODE REVIEW

File: conda_build/api.py [hunk 1]
```diff
@@ -395,8 +395,8 @@ def update_index(dir_paths, config=None, force=False, check_md5=False, remove=Fa
                      subdirs=ensure_list(subdir), convert_if_not_present=convert_if_not_present)
 
 
-def debug(recipe_or_package_path_or_metadata_tuples, path=None, test=False, output_id=None, config=None,
-          verbose=True, **kwargs):
+def debug(recipe_or_package_path_or_metadata_tuples, path=None, test=False,
+          link_source_method='auto', output_id=None, config=None, verbose=True, **kwargs):
```
Comment (msarahan) [inline conda_build/api.py]: When adding new arguments like this, it's better to put them at the end (here, between verbose=True and **kwargs).  Keyword arguments can still be passed positionally, and we should avoid accidentally breaking people who choose to only do positional args (crazy though they may be).

File: conda_build/post.py [hunk 2]
```diff
@@ -427,6 +429,11 @@ def mk_relative_linux(f, prefix, rpaths=('lib',)):
     except CalledProcessError:
         print('patchelf: --print-rpath failed for %s\n' % (elf))
         return
+    print("existing paths via patchelf: {}".format(existing))
```
Comment (msarahan) [inline conda_build/post.py]: This looks like debugging output.  Can it be done with log.debug instead of print?

**Comment by msarahan** [comment]
This looks mostly fine to me.  A few requests:

1. fix flake8 and the merge conflict, please
2. there's a failure with unicode_all_over on py27.  If this is expected/acceptable, can we just xfail this test?
3. we need to understand what the slowdown is here.  I don't see anything super obvious, but this PR seems to have done something that slows things down a lot.  I can help with profiling if you need it.

**Comment by nehaljwani** [comment]
py27 failure on Linux:
```
tests/test_api_build.py::test_recipe_builds[unicode_all_over] FAILED                                                                                                                                 [100%]

================================================================================================= FAILURES =================================================================================================
___________________________________________________________________________________ test_recipe_builds[unicode_all_over] ___________________________________________________________________________________
Traceback (most recent call last):
  File "~/repos/conda-build/tests/test_api_build.py", line 96, in test_recipe_builds
    api.build(recipe, config=testing_config)
  File "~/repos/conda-build/conda_build/api.py", line 209, in build
    notest=notest, need_source_download=need_source_download, variants=variants)
  File "~/repos/conda-build/conda_build/build.py", line 2282, in build_tree
    notest=notest,
  File "~/repos/conda-build/conda_build/build.py", line 1594, in build
    newly_built_packages = bundlers[pkg_type](output_d, m, env, stats)
  File "~/repos/conda-build/conda_build/build.py", line 983, in bundle_conda
    basename + ext, out_folder=tmp)
  File "~/ccc/lib/python2.7/site-packages/conda_package_handling/api.py", line 46, in create
    SUPPORTED_EXTENSIONS[ext].create(prefix, file_list, out_fn, out_folder, **kw)
  File "~/ccc/lib/python2.7/site-packages/conda_package_handling/tarball.py", line 102, in create
    '.tar.bz2', 'bzip2')
  File "~/ccc/lib/python2.7/site-packages/conda_package_handling/tarball.py", line 64, in create_compressed_tarball
    options=filter_opts) as archive:
  File "~/ccc/lib/python2.7/contextlib.py", line 17, in __enter__
    return self.gen.next()
  File "~/ccc/lib/python2.7/site-packages/libarchive/write.py", line 175, in file_writer
    ffi.write_open_filename_w(archive_p, filepath)
  File "~/ccc/lib/python2.7/site-packages/libarchive/ffi.py", line 91, in check_int
    raise archive_error(args[0], retcode)
  File "~/ccc/lib/python2.7/site-packages/libarchive/ffi.py", line 75, in archive_error
    raise ArchiveError(msg, errno(archive_p), retcode, archive_p)
ArchiveError: Failed to clean up compressor (errno=22, retcode=-30, archive_p=94693808656128)
------------------------------------------------------------------------------------------- Captured stdout call ------------------------------------------------------------------------------------------$
Adding in variants from internal_defaults
Attempting to finalize metadata for ♥
BUILD START: [u'\u2665-1-0.tar.bz2']
```

**Comment by nehaljwani** [comment]
Another py27 failure on linux:
```

tests/test_api_build.py::test_only_lua_env FAILED                                                                                                                                                    [100%]

================================================================================================= FAILURES =================================================================================================
____________________________________________________________________________________________ test_only_lua_env _____________________________________________________________________________________________
Traceback (most recent call last):
  File "/apps/infrafs1/nwani/repos/conda-build/tests/test_api_build.py", line 1066, in test_only_lua_env
    api.build(recipe, config=testing_config)
  File "/apps/infrafs1/nwani/repos/conda-build/conda_build/api.py", line 209, in build
    notest=notest, need_source_download=need_source_download, variants=variants)
  File "/apps/infrafs1/nwani/repos/conda-build/conda_build/build.py", line 2282, in build_tree
    notest=notest,
  File "/apps/infrafs1/nwani/repos/conda-build/conda_build/build.py", line 1594, in build
    newly_built_packages = bundlers[pkg_type](output_d, m, env, stats)
  File "/apps/infrafs1/nwani/repos/conda-build/conda_build/build.py", line 957, in bundle_conda
    files = post_process_files(metadata, initial_files)
  File "/apps/infrafs1/nwani/repos/conda-build/conda_build/build.py", line 824, in post_process_files
    post_build(m, new_files, build_python=python)
  File "/apps/infrafs1/nwani/repos/conda-build/conda_build/post.py", line 1063, in post_build
    check_overlinking(m, files)
  File "/apps/infrafs1/nwani/repos/conda-build/conda_build/post.py", line 1002, in check_overlinking
    m.config.channel_urls)
  File "/apps/infrafs1/nwani/repos/conda-build/conda_build/post.py", line 916, in check_overlinking_impl
    files, run_prefix, build_prefix, all_needed_dsos, pkg_vendored_dist, ignore_list_syms, sysroot_substitution)
  File "/apps/infrafs1/nwani/repos/conda-build/conda_build/post.py", line 666, in _map_file_to_package
    exports = set(e for e in get_exports_memoized(fp) if not
  File "/apps/infrafs1/nwani/repos/conda-build/conda_build/os_utils/liefldd.py", line 849, in __call__
    value = self.func(*args, **kw)
  File "/apps/infrafs1/nwani/repos/conda-build/conda_build/os_utils/liefldd.py", line 856, in get_exports_memoized
    return get_exports(filename, arch=arch)
  File "/apps/infrafs1/nwani/repos/conda-build/conda_build/os_utils/liefldd.py", line 720, in get_exports
    exports2, flags2, exports2_all, flags2_all = get_static_lib_exports(filename)
  File "/apps/infrafs1/nwani/repos/conda-build/conda_build/os_utils/liefldd.py", line 667, in get_static_lib_exports
    obj = lief.ELF.parse(raw=content[obj_start:obj_end])
TypeError: parse(): incompatible function arguments. The following argument types are supported:
    1. (filename: unicode, dynsym_count_method: _pylief.ELF.DYNSYM_COUNT_METHODS=DYNSYM_COUNT_METHODS.AUTO) -> LIEF::ELF::Binary
    2. (raw: List[int], name: unicode=u'', dynsym_count_method: _pylief.ELF.DYNSYM_COUNT_METHODS=DYNSYM_COUNT_METHODS.AUTO) -> LIEF::ELF::Binary
    3. (io: object, name: unicode=u'') -> LIEF::ELF::Binary

Invoked with: kwargs: raw='\x7fELF\x02\x01\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01\x00>\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x88\x06\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00@\x00\x00\x00\x00\x00@\x00\x0c\x00\x0b\x00USH\x89\xfbH\x83\xec\x08H\x8b\xbf`\x05\x00\x00\x8b\x83h\x05\x00\x00\x0f\xbfktH\x85\xff\x0f\x
84\x8d\x00\x00\x009\xe8\x0f\x8c\x85\x00\x00\x00H\x89\xdf1\xed\xff\x15\x00\x00\x00\x00D\x0f\xbfCtE\x85\xc0~-H\x8b\xbb`\x05\x00\x00Hc\xc5\x89\xeaH\x8d\x04\x87\x8b0\x83\xc5\x01\x83\xfe\xfft\x089\xd6\x0f\x8f\xfe\x00\x00\x00H\x83\xc0\x04\x89\xeaA9\xe8\x7f\xe3D\x89\xc1\x83\xe9\x01x4\x0f\x1f\x00L\x8b\x83`\x05\x00\x00Hc\xc1I\x8d\x04\x80f.\x0f\x1f\x84\x00\x00\x00\
x00\x00\x8b\x10\x8di\xff9\xca}\x05\x83\xfa\xffuBH\x83\xe8\x04\x89\xe9\x83\xfd\xffu\xe7X[]\xc3\x0f\x1f\x009\xe8\x0fM\xe8Hc\xf5H\xc1\xe6\x02\xff\x15\x00\x00\x00\x00H\x85\xc0t\xe2H\x89\x83`\x05\x00\x00\x89\xabh\x05\x00\x00\xe9R\xff\xff\xff\x0f\x1f\x80\x00\x00\x00\x00\x89\xd6)\xce\x83\xfd\xfftZHc\xc5H\x8d<\x85\x00\x00\x00\x00A\x8b\x04\x80\x83\xf8\xff\x0f\x84\
x1f\x01\x00\x00)\xe89\xc6\x0f\x85\x15\x01\x00\x00\x8dA\xfeH\x98H\x89\xc2H\xf7\xdaH\x8d<\x97L\x01\xc7\xeb\x1c\x90\x8bT\x87\xfc\x83\xfa\xff\x0f\x84\xc3\x00\x00\x00)\xeaH\x83\xe8\x019\xd6\x0f\x85\xb5\x00\x00\x00\x89\xc5H\x83\xf8\xffu\xddD\x0f\xbfCtH\x83\xc4\x08H\x89\xdf\x89\xf2[]A\x83\xe8\x01\xff%\x00\x00\x00\x00\x0f\x1f\x00)\xd6A9\xe8\x0f\x8e\xad\x00\x00\x0
0Hc\xc5L\x8d\x0c\x85\x00\x00\x00\x00\x8b\x04\x87\x83\xf8\xff\x0f\x84\x96\x00\x00\x00)\xe89\xc6\x0f\x85\x8c\x00\x00\x00\x8dB\x02A\x83\xe8\x01H\x98D\x89\xc1I\x89\xc2)\xe9I\xf7\xdaH\x01\xc1O\x8d\x0c\x91I\x01\xf9\xeb\x18\x0f\x1f@\x00A\x8b|\x81\x04\x83\xff\xfft\x11)\xefH\x83\xc0\x019\xfeu\x07\x89\xc5H9\xc1u\xe5\x8dL.\xffH\x89\xdf\xff\x15\x00\x00\x00\x00D\x0f\x
bfCtA9\xe8\x0f\x8fZ\xfe\xff\xff\xe9\x82\xfe\xff\xff\x0f\x1fD\x00\x00\x0f\xbfCt\x8dT.\x01H\x89\xdfD\x8d@\xff\xff\x15\x00\x00\x00\x00\x83\xfd\xff\x0f\x84\x9b\xfe\xff\xff\x89\xe9\xe9c\xfe\xff\xff\x0f\x1f\x00A\x83\xe8\x01\xeb\xadf\x90D\x0f\xbfCtH\x89\xdfA\x83\xe8\x01\xff\x15\x00\x00\x00\x00\x89\xe9\xe9?\xfe\xff\xff\x0f\x1f\x80\x00\x00\x00\x00H\x8b\x05\x00\x00
\x00\x00H\x8b8\xff%\x00\x00\x00\x00\x00GCC: (crosstool-NG 1.23.0.449-a04d0) 7.3.0\x00\x00\x00\x00\x00\x14\x00\x00\x00\x00\x00\x00\x00\x01zR\x00\x01x\x10\x01\x1b\x0c\x07\x08\x90\x01\x00\x004\x00\x00\x00\x1c\x00\x00\x00\x00\x00\x00\x009\x02\x00\x00\x00A\x0e\x10\x86\x02A\x0e\x18\x83\x03G\x0e \x02\xa1\n\x0e\x18A\x0e\x10A\x0e\x08D\x0b\x02\x9c\n\x0e\x18F\x0e\x1
0A\x0e\x08M\x0b\x14\x00\x00\x00T\x00\x00\x00\x00\x00\x00\x00\x10\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x04\x00\xf1\xff\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x03\x00\x01\x00\x00\x00\x00\x00\x00\x00\
x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x03\x00\x03\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x03\x00\x04\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x03\x00\x06\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x03\x00\x07\x
00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x03\x00\x05\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x0e\x00\x00\x00\x12\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00\x009\x02\x00\x00\x00\x00\x00\x00%\x00\x00\x00\x10\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00;\x00\x00
\x00\x10\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00K\x00\x00\x00\x10\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00W\x00\x00\x00\x10\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00f\x00\x00\x00\x12\x00\x01\x00@\x02\x00\x00\x00\x00\x00\x00\x10\x00\x00\x00\x00\x00\x00\
x00z\x00\x00\x00\x10\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00}\x00\x00\x00\x11\x00\x04\x00\x00\x00\x00\x00\x00\x00\x00\x00\x08\x00\x00\x00\x00\x00\x00\x00\x00hardscroll.c\x00_nc_scroll_optimize_sp\x00_GLOBAL_OFFSET_TABLE_\x00_nc_hash_map_sp\x00_nc_doalloc\x00_nc_scrolln_sp\x00_nc_scroll_optimize\x00SP\x00_nc_oldnums\x00\
x00\x00\x00\x00\x00\x00\x002\x00\x00\x00\x00\x00\x00\x00)\x00\x00\x00\n\x00\x00\x00\xfc\xff\xff\xff\xff\xff\xff\xff\xbe\x00\x00\x00\x00\x00\x00\x00)\x00\x00\x00\x0b\x00\x00\x00\xfc\xff\xff\xff\xff\xff\xff\xffY\x01\x00\x00\x00\x00\x00\x00)\x00\x00\x00\x0c\x00\x00\x00\xfc\xff\xff\xff\xff\xff\xff\xff\xd4\x01\x00\x00\x00\x00\x00\x00)\x00\x00\x00\x0c\x00\x00\x
00\xfc\xff\xff\xff\xff\xff\xff\xff\x01\x02\x00\x00\x00\x00\x00\x00)\x00\x00\x00\x0c\x00\x00\x00\xfc\xff\xff\xff\xff\xff\xff\xff.\x02\x00\x00\x00\x00\x00\x00)\x00\x00\x00\x0c\x00\x00\x00\xfc\xff\xff\xff\xff\xff\xff\xffC\x02\x00\x00\x00\x00\x00\x00*\x00\x00\x00\x0e\x00\x00\x00\xfc\xff\xff\xff\xff\xff\xff\xffL\x02\x00\x00\x00\x00\x00\x00)\x00\x00\x00\x08\x00
\x00\x00\xfc\xff\xff\xff\xff\xff\xff\xff \x00\x00\x00\x00\x00\x00\x00\x02\x00\x00\x00\x02\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00X\x00\x00\x00\x00\x00\x00\x00\x02\x00\x00\x00\x02\x00\x00\x00@\x02\x00\x00\x00\x00\x00\x00\x00.symtab\x00.strtab\x00.shstrtab\x00.rela.text\x00.data\x00.bss\x00.comment\x00.note.GNU-stack\x00.rela.eh_frame\x00\x00\x00\x00\x0
0\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00 \x00\x00\x00\x01\x00\x00\x00\x06\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x0
0\x00@\x00\x00\x00\x00\x00\x00\x00P\x02\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x10\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x1b\x00\x00\x00\x04\x00\x00\x00@\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x008\x05\x00\x00\x00\x00\x00\x00\xc0\x00\x00\x00\x00\x00\x00\x00\t\x00\x00\x00\x01\x00\x00\x00\x08\x00\x00\x
00\x00\x00\x00\x00\x18\x00\x00\x00\x00\x00\x00\x00&\x00\x00\x00\x01\x00\x00\x00\x03\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x90\x02\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00,\x00\x00\x00\x08\x00\x00\x00\x03\x00\x00\x00\x00\x00\
x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x90\x02\x00\x00\x00\x00\x00\x00\x08\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x08\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x001\x00\x00\x00\x01\x00\x00\x000\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x90\x02\x00\x00\x00\x00\x00\x00,\x00\x00\x00\x00\x00\x00\x00\x00\x0
0\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00:\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xbc\x02\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00O\x00\x00\x00\x
01\x00\x00\x00\x02\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xc0\x02\x00\x00\x00\x00\x00\x00h\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x08\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00J\x00\x00\x00\x04\x00\x00\x00@\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xf8\x05\x00\x00\x00\x00\x00\x00
0\x00\x00\x00\x00\x00\x00\x00\t\x00\x00\x00\x07\x00\x00\x00\x08\x00\x00\x00\x00\x00\x00\x00\x18\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x02\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00(\x03\x00\x00\x00\x00\x00\x00\x80\x01\x00\x00\x00\x00\x00\x00\n\x00\x00\x00\x08\x00\x00\x00\x08\x00\x00\x00\x00\x00\x00\x00\x18\x00\x00\x0
0\x00\x00\x00\x00\t\x00\x00\x00\x03\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xa8\x04\x00\x00\x00\x00\x00\x00\x89\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x11\x00\x00\x00\x03\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x
00\x00\x00(\x06\x00\x00\x00\x00\x00\x00Y\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'
```

**Comment by nehaljwani** [comment]
Let's continue fixing tests and profiling in further PRs.

**Pull request merged**